### PR TITLE
SW-355: WPOverlay -  Adjust the way DO_JUMP retraces the route

### PR DESCRIPTION
**Pull request by LachlanConn**
When a DO_JUMP command is used mission planner retraces the route to allow for the progress bar to accurately report how much longer is left in a planned flight. During this retrace there is no consideration for specific rules in drawing this route such as Loiter exit tangent commands. This pull request aims to fix this but there are some edge cases where if a do jump is commanded after a VTOL_LAND and references a WP before the VTOL land it will not correctly draw the route but this is a situation that should never really happen.
Before: 
image
After:
image

### CODE REVIEW

**Comment by robertlong13** [comment]
Can you share a screenshot of the VTOL Land edge case that you mentioned?

This looks great btw

**Comment by LachlanConn** [comment]
This is when it references a wp after a VTOL_LAND correctly retracing the Route.
image
This is when referencing a wp before the VTOL_LAND (it just doesn't retrace anything)
image
This is what base mission planner does 
image

Basically when it gets to a VTOL_LAND it clears the route so it can't retrace correctly. I have introduced this because now I actually check the conditions of the waypoints when retracing the routes.

File: ExtLibs/Maps/WPOverlay.cs [hunk 1]
```diff
@@ -88,31 +92,28 @@ public void CreateOverlay(PointLatLngAlt home, List<Locationwp> missionitems, do
...
-                        route.Add(pointlist[pointlist.Count - 1]);
+                        AddPointToRoute(a + 1, pointlist[pointlist.Count - 1]);
 
                         dolandstart = a;
                         // draw everything before
                         if (route.Count > 0)
                         {
-                            RegenerateWPRoute(route, home, false);
-                            route.Clear();
+                            pointlist[pointlist.Count - 1].Tag = "break";
```
Comment (robertlong13) [inline ExtLibs/Maps/WPOverlay.cs:101]: This overwrites the "DLS" tag I'm adding here, which breaks the elevation profile for landing sequences with hold times. I know we don't use it right now, but let's not regress it. Should be able to just delete this bit and add a `Contains("DLS")` check later.

File: ExtLibs/Maps/WPOverlay.cs [hunk 2]
```diff
@@ -574,8 +582,19 @@ private void RegenerateWPRoute(List<PointLatLngAlt> fullpointlist, PointLatLngAl
                         }
                         return;
                     }
+                    if (x.Tag == "break")
```
Comment (robertlong13) [inline ExtLibs/Maps/WPOverlay.cs:585]:
```suggestion
                    if (x.Tag == "break" || x.Tag.Contains("DLS"))
```

**Review by robertlong13 — CHANGES_REQUESTED** [review]
I'm good to merge this with one minor change.

File: ExtLibs/Maps/WPOverlay.cs [hunk 3]
```diff
@@ -259,19 +255,30 @@ public void CreateOverlay(PointLatLngAlt home, List<Locationwp> missionitems, do
 
                     int wpno = (int) Math.Max(item.p1, 0);
                     int repeat = (int)item.p2;
+                    if (wpno > a)
```
Comment (robertlong13) [inline ExtLibs/Maps/WPOverlay.cs:258]: This needs a lengthy, borderline apologetic, downright self-deprecating explanation of why this is needed.

Comment (LachlanConn) [reply ExtLibs/Maps/WPOverlay.cs:262]: Let me know if its apologetic enough or if I need to be more self deprecating

**Review by robertlong13 — APPROVED** [review]
I force pushed a whitespace nitpick ([space between comment marker and start of comment](https://stackoverflow.com/questions/1467058/space-between-line-comment-characters-and-start-of-actual-comment)), and a commit message nitpick.

Good shit though, Lachie. Merging

**Pull request merged**